NorthstarDLL/server/auth/bansystem.cpp

```diff
@@ -213,12 +216,59 @@ void ConCommand_clearbanlist(const CCommand& args)
 	g_pBanSystem->ClearBanlist();
 }
 
+int ConCommand_banCompletion(const char* const partial, char commands[COMMAND_COMPLETION_MAXITEMS][COMMAND_COMPLETION_ITEM_LENGTH])
+{
+	const char* space = strchr(partial, ' ');
+	const char* cmdName = partial;
+	const char* query = partial + (space == nullptr ? 0 : space - partial) + 1;
+
+	const int queryLength = strlen(query);
+	const int cmdLength = strlen(cmdName) - queryLength;
+
+	int numCompletions = 0;
+	for (int i = 0; i < R2::GetMaxPlayers() && numCompletions < COMMAND_COMPLETION_MAXITEMS - 2; i++)
+	{
+		R2::CBaseClient* client = &R2::g_pClientArray[i];
+		if (client->m_Signon < R2::eSignonState::CONNECTED)
+			continue;
+
+		if (!strncmp(query, client->m_Name, queryLength))
+		{
+			strncpy(commands[numCompletions], cmdName, cmdLength);
+			strncpy_s(
+				commands[numCompletions++] + cmdLength,
+				COMMAND_COMPLETION_ITEM_LENGTH,
+				client->m_Name,
+				COMMAND_COMPLETION_ITEM_LENGTH - cmdLength);
+		}
+
+		if (!strncmp(query, client->m_UID, queryLength))
+		{
+			strncpy(commands[numCompletions], cmdName, cmdLength);
+			strncpy_s(
+				commands[numCompletions++] + cmdLength,
+				COMMAND_COMPLETION_ITEM_LENGTH,
+				client->m_UID,
+				COMMAND_COMPLETION_ITEM_LENGTH - cmdLength);
+		}
+	}
+
+	return numCompletions;
+}
+
 ON_DLL_LOAD_RELIESON("engine.dll", BanSystem, ConCommand, (CModule module))
 {
 	g_pBanSystem = new ServerBanSystem;
 	g_pBanSystem->OpenBanlist();
 
-	RegisterConCommand("ban", ConCommand_ban, "bans a given player by uid or name", FCVAR_GAMEDLL);
-	RegisterConCommand("unban", ConCommand_unban, "unbans a given player by uid", FCVAR_GAMEDLL);
+	RegisterConCommand("ban", ConCommand_ban, "bans a given player by uid or name", FCVAR_GAMEDLL, ConCommand_banCompletion);
+	RegisterConCommand("unban", ConCommand_unban, "unbans a given player by uid", FCVAR_GAMEDLL, ConCommand_banCompletion);
 	RegisterConCommand("clearbanlist", ConCommand_clearbanlist, "clears all uids on the banlist", FCVAR_GAMEDLL);
 }
+
+ON_DLL_LOAD_RELIESON("server.dll", KickCompletion, ConCommand, (CModule module))
+{
+	ConCommand* kick = R2::g_pCVar->FindCommand("kick");
```

Comment:
@catornot builds are failing atm as `R2` namespace was removed

Reply:
removed it